src/io_callbacks.zig

```diff
@@ -0,0 +1,463 @@
+const std = @import("std");
+const assert = std.debug.assert;
+const os = std.os;
+const linux = os.linux;
+const IO_Uring = linux.IO_Uring;
+const io_uring_cqe = linux.io_uring_cqe;
+const io_uring_sqe = linux.io_uring_sqe;
+
+pub const IO = struct {
+    /// An intrusive first in/first out linked list.
+    const Fifo = struct {
+        in: ?*Completion = null,
+        out: ?*Completion = null,
+
+        fn push(fifo: *Fifo, completion: *Completion) void {
+            assert(completion.next == null);
+            if (fifo.in) |in| {
+                in.next = completion;
+                fifo.in = completion;
+            } else {
+                assert(fifo.out == null);
+                fifo.in = completion;
+                fifo.out = completion;
+            }
+        }
+    };
+
+    ring: IO_Uring,
+
+    /// Completions not yet submitted to the kernel and waiting on available space in the
+    /// submission queue.
+    unqueued: Fifo = .{},
+
+    /// The number of SQEs queued but not yet submitted to the kernel:
+    queued: u32 = 0,
+
+    /// The number of SQEs submitted and inflight but not yet completed.
+    submitted: u32 = 0,
+
+    /// Completions that are ready to have their callbacks run.
+    completed: Fifo = .{},
+
+    pub fn init(entries: u12, flags: u32) !IO {
+        return IO{ .ring = try IO_Uring.init(entries, flags) };
+    }
+
+    pub fn deinit(self: *IO) void {
+        self.ring.deinit();
+    }
+
+    /// Pass all queued submissions to the kernel and run the event loop
+    /// until there is no longer any i/o pending.
+    pub fn run(self: *IO) !void {
+        while (self.queued + self.submitted > 0 or
+            self.unqueued.out != null or self.completed.out != null)
```

Comment:
If this was Haiku then it would be cool to reuse the completed queue to handle a full submission queue, then we can make this a one-liner, but I think the way you've got it is the simplest already.

Reply:
Thinking about this a bit more I don't think we actually need the `completed` FIFO since we are using callbacks not async. Do you see a reason that would stop us from getting rid of it and just executing the callbacks directly in `flush_completions()`?

Reply:
Great, I was hoping this would just be food for thought. Let me take a look and see...